test/jdk/tools/jpackage/windows/WinNoRestartTest.java

```diff
@@ -0,0 +1,154 @@
+ * @library ../helpers
+ * @library /test/lib
+ * @requires os.family == "windows"
+ * @build jdk.jpackage.test.*
+ * @build WinNoRestartTest
+ * @modules jdk.jpackage/jdk.jpackage.internal
+ * @run main/othervm -Xmx512m jdk.jpackage.test.Main
+ *  --jpt-run=WinNoRestartTest
+ *
+ */
+import java.io.IOException;
+import java.time.Duration;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import jdk.jpackage.test.JPackageCommand;
+import jdk.jpackage.test.Annotations.Test;
+import jdk.jpackage.test.CfgFile;
+import jdk.jpackage.test.HelloApp;
+import static jdk.jpackage.test.WindowsHelper.findAppLauncherPID;
+import static jdk.jpackage.test.WindowsHelper.killProcess;
+
+public class WinNoRestartTest {
+
+    @Test
+    public static void test() throws InterruptedException, IOException {
+        var cmd = JPackageCommand.helloAppImage().ignoreFakeRuntime();
+
+        // Configure test app to launch in a way it will not exit
+        cmd.addArguments("--java-options", "-Djpackage.test.noexit=true");
+        cmd.execute();
+
+        var origCfgFile = CfgFile.load(cmd.appLauncherCfgPath(null));
+
+        for (var testConfig : testConfig()) {
+            testConfig.apply(cmd, origCfgFile);
+        }
+    }
+
+    private static record NoRerunSectionConfig(String sectionName,
+            String firstValue, String secondValue) {
+
+        void apply(CfgFile cfgFile) {
+            if (firstValue != null) {
+                cfgFile.addValue(sectionName, "win.norestart", firstValue);
+            }
+            if (secondValue != null) {
+                cfgFile.addValue(sectionName, "win.norestart", secondValue);
+            }
+        }
+    }
+
+    private static record NoRerunConfig(NoRerunSectionConfig firstSection,
+            NoRerunSectionConfig secondSection, boolean expectedNoRestarted) {
+
+        void apply(JPackageCommand cmd, CfgFile origCfgFile) throws InterruptedException {
+            // Alter the main launcher .cfg file
+            var cfgFile = new CfgFile();
+            if (firstSection != null) {
+                firstSection.apply(cfgFile);
+            }
+            cfgFile = CfgFile.combine(cfgFile, origCfgFile);
+
+            if (secondSection != null) {
+                secondSection.apply(cfgFile);
+            }
+
+            // Save updated main launcher .cfg file
+            cfgFile.save(cmd.appLauncherCfgPath(null));
+
+            try ( // Launch the app in a separate thread
+                ExecutorService exec = Executors.newSingleThreadExecutor()) {
+                exec.execute(() -> {
+                    HelloApp.executeLauncher(cmd);
+                });
+
+                // Wait a bit to let the app start
+                Thread.sleep(Duration.ofSeconds(10));
+
+                // Get PID of the main app launcher process
+                final long pid = findAppLauncherPID(cmd, null,
```

Comment:
If I am reading code correctly `TKit.assertEquals` will throw exception, which means `killProcess` might not be called in case if we expecting 1 process, but we got 2.

Reply:
Good catch! I didn't think through it well. Will rework